drivers/kscan/kscan_matrix_poll.c

```diff
@@ -0,0 +1,303 @@
+			 */
+			kb_poll.matrix_stable_state[c] ^= mask;
+			kb_poll.config.callback(dev, r, c, row_bit ? true : false);
+		}
+	}
+
+	return key_pressed;
+}
+
+void polling_task(void *dummy1, void *dummy2, void *dummy3)
+{
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period;
+
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (true) {
+		kscan_resume_matrix_detection(kscan_dev, true);
+		kscan_drive_matrix_column(kscan_dev, KEYBOARD_COLUMN_DRIVE_ALL);
+		k_sem_take(&poll_lock, K_FOREVER);
+		LOG_DBG("Start KB scan!!");
+
+		/* Start polling */
+		kscan_resume_matrix_detection(kscan_dev, false);
+
+		uint64_t poll_time_end = sys_clock_timeout_end_calc(K_USEC(kb_poll.poll_timeout));
+
+		while (atomic_get(kb_poll.config.enable_scan) == 1U) {
+			uint32_t start_period_cycles = k_cycle_get_32();
+
+			if (check_key_events(kscan_dev)) {
+				poll_time_end =
+					sys_clock_timeout_end_calc(K_USEC(kb_poll.poll_timeout));
+
+			} else if (k_uptime_ticks() > poll_time_end) {
+				break;
+			}
+
+			/* Subtract the time invested from the sleep period
+			 * in order to compensate for the time invested
+			 * in debouncing a key
+			 */
+			current_cycles = k_cycle_get_32();
+			cycles_diff = current_cycles - start_period_cycles;
+			wait_period = kb_poll.poll_period - k_cyc_to_us_floor32(cycles_diff);
+
+			/* Override wait_period in case it is less than 1 ms */
+			if (wait_period < USEC_PER_MSEC) {
+				wait_period = USEC_PER_MSEC;
+			}
+
+			/* wait period results in a larger number when
+			 * current cycles counter wrap. In this case, the
+			 * whole poll period is used
+			 */
+			if (wait_period > kb_poll.poll_period) {
+				LOG_DBG("wait_period : %u", wait_period);
+
+				wait_period = kb_poll.poll_period;
+			}
+
+			/* Allow other threads to run while we sleep */
+			k_usleep(wait_period);
+		}
+	}
+}
+
+static void kscan_ksi_isr_callback(const struct device *dev)
+{
+	kscan_resume_matrix_detection(kscan_dev, false);
+	k_sem_give(&poll_lock);
+}
+
+int kscan_matrix_poll_task_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	kscan_dev = device_get_binding(KSCAN_DEV_NAME);
+	/* Initialize semaphore usbed by kscan task and driver */
+	k_sem_init(&poll_lock, 0, 1);
+	kb_poll.config.isr_callback = kscan_ksi_isr_callback;
+
+	/* Time figures are transformed from msec to usec */
+	kb_poll.deb_time_press = (uint32_t)(CONFIG_KSCAN_MATRIX_DEBOUNCE_DOWN * USEC_PER_MSEC);
+	kb_poll.deb_time_rel = (uint32_t)(CONFIG_KSCAN_MATRIX_DEBOUNCE_UP * USEC_PER_MSEC);
+	kb_poll.poll_period = (uint32_t)(CONFIG_KSCAN_MATRIX_POLL_PERIOD * USEC_PER_MSEC);
+	kb_poll.poll_timeout = 100 * USEC_PER_MSEC;
+
+	kscan_config_matrix(kscan_dev, &kb_poll.config);
+
+	k_thread_create(&kb_poll.thread, kb_poll.thread_stack, TASK_STACK_SIZE, polling_task, NULL,
+			NULL, NULL, K_PRIO_COOP(4), 0, K_NO_WAIT);
+
+	return 0;
+}
+
+SYS_INIT(kscan_matrix_poll_task_init, POST_KERNEL, CONFIG_KSCAN_MATRIX_TASK_INIT_PRIORITY);
```

Comment:
this is not a driver, so I'm not sure its place is `drivers/kscan`. I wonder why this isn't provided as an API enhancement, a sort of non-device specific layer on top of the API.